### PR TITLE
Refs #28643 -- Added SHA1, SHA224, SHA256, SHA384 and SHA512 database functions.

### PR DESCRIPTION
**Pull request by pope1ni**
Ticket [#28643](https://code.djangoproject.com/ticket/28643).

I was inspired by #11004 :grin: This is not perfect and I expect a few things will blow up initially...

### CODE REVIEW

**Comment by pope1ni** [comment]
buildbot, test on oracle.

File: django/db/models/functions/text.py [hunk 1]
```diff
@@ -219,6 +219,60 @@ class RTrim(Transform):
...
+        return super().as_sql(
+            compiler,
+            connection,
+            template="ENCODE(DIGEST((%(expressions)s), '%(function)s'), 'hex')",
+            function=self.function.lower(),
+            **extra_content,
+        )
+
+
+class SHA224(SHA1):  # FIXME: Not supported on Oracle?
```
Comment (pope1ni) [inline django/db/models/functions/text.py]: Looks like Oracle is the only one not to support this... Not sure how we'd handle that...

File: tests/db_functions/migrations/0001_setup_extensions.py [hunk 2]
```diff
@@ -0,0 +1,15 @@
+from unittest import mock
+
+from django.db import migrations
+
+try:
+    from django.contrib.postgres.operations import CryptoExtension
```
Comment (pope1ni) [inline tests/db_functions/migrations/0001_setup_extensions.py]: This sucks, but not sure how else we can satisfy this for the tests...

File: tests/db_functions/migrations/0001_setup_extensions.py [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+except ImportError:
+    CryptoExtension = mock.Mock()
+
+
+class Migration(migrations.Migration):
+
+    # XXX: Enable pg_crypto extension on PostgreSQL to support various database
+    #      functions including SHA1, SHA224, ShA256, SHA384 & SHA512.
```
Comment (pope1ni) [inline tests/db_functions/migrations/0001_setup_extensions.py]: Urgh. For my own reminder, I typoed here...

**Comment by felixxm** [comment]
@pope1ni Thanks for this patch :+1: To be honest I added PR #11004 to finally close this huge [ticket](https://code.djangoproject.com/ticket/28643) :wink: . I would prefer to add separate ticket for adding other hash methods because adding them is more complicated than supporting `MD5` e.g. some of them are not supported on all databases, `pgcrypto` extension is required on PostgreSQL etc. What do you think?

**Comment by pope1ni** [comment]
buildbot, test on oracle.

**Comment by pope1ni** [comment]
> ...to finally close this huge [ticket](https://code.djangoproject.com/ticket/28643)...
> What do you think?

Seems reasonable. I'll look into creating another ticket later. Have just pushed a bunch of tweaks to fix the obvious and update to match the current state of your work on `MD5()`.

File: django/db/models/functions/text.py (hunk 1 above)
Comment (pope1ni) [reply django/db/models/functions/text.py]: As expected...
```
  test_basic (db_functions.text.test_sha224.SHA224Tests) ... ERROR (0.012s)
  test_transform (db_functions.text.test_sha224.SHA224Tests) ... ERROR (0.006s)
```

Comment (felixxm) [reply django/db/models/functions/text.py]: It seems that there is no way to emulate `SHA224` on Oracle, even with [`DBMS_CRYPTO`](https://docs.oracle.com/en/database/oracle/oracle-database/18/arpls/DBMS_CRYPTO.html#GUID-CE3CF17D-E781-47CB-AEE7-19A9B2BCD3EC) :disappointed: .

**Comment by pope1ni** [comment]
@felixxm I opened a new ticket [#30240](https://code.djangoproject.com/ticket/30240) and have created a new PR #11059 with a few additional tweaks.

**Pull request closed without merging**